File: prog/hw5/R/hw5.R
#!/usr/bin/Rscript


### Assignment 5 ===============================================================
library(quadprog)            # attach package quadprog for quadratic programming


## Question 1. Tangency Portfolio ----------------------------------------------

## same setting from example code
na <- 20L                               # number of assets
ns <- 60L                               # number of observations
set.seed(1216)                          # important for reproducibility
retn <- matrix(rnorm(ns * na, mean = 0.005, sd = 0.015), nrow = ns)
mu <- colMeans(retn)                    # expected returns
rf <- 0.0001                            # riskfree rate (about 2.5% pa)
mu2 <- mu - rf                          # excess means
Q <- cov(retn)                          # covariance matrix

## obtain the optimized weights by quadratic programming
res_tan <- solve.QP(Dmat = Q, dvec = rep(0, na), Amat = matrix(mu2, ncol = 1),
                    bvec = 1, meq  = 1L)
(w_tan <- prop.table(res_tan$solution)) # rescale variables to obtain weights
(retn_tan <- crossprod(w_tan, mu))      # expected return
(se_tan <- as.vector(sqrt(crossprod(w_tan, Q %*% w_tan)))) # standard error
(sr <- crossprod(w_tan, mu2) / se_tan)  # compute sharpe ratio
stopifnot(all.equal(sum(w_tan), 1))     # check budget constraint
stopifnot(all(retn_tan >= rf))          # check return constraint

## minimal variance portfolio
res_min <- solve.QP(Dmat = 2 * Q, dvec = rep(0, na),
                    Amat = matrix(1, nrow = na, ncol = 1),
                    bvec = 1, meq = 1L)
(w_min <- res_min$solution)
(retn_min <- crossprod(w_min, mu))      # expected return
(se_min <- sqrt(
     var_min <- as.vector(crossprod(w_min, Q %*% w_min)))) # standard error
stopifnot(all.equal(sum(w_min), 1))
stopifnot(all.equal(var_min, res_min$value))

## construct efficient frontier from minimal variance and tangency portfolio
retnVec <- c(retn_min, retn_tan)
seVec <- c(se_min, se_tan)
varMat <- diag(seVec ^ 2)
varMat[1, 2] <- varMat[2, 1] <- as.vector(crossprod(w_min, Q %*% w_tan))
alpha <- seq(from = 1, to = - 1, by = - 0.1)
nAlpha <- length(alpha)
tmpMat <- cbind(alpha, 1 - alpha)
retn_eff <- tmpMat %*% retnVec
var_eff <- sapply(alpha, function(a) {
    crossprod(c(a, 1 - a), varMat %*% c(a, 1 - a))
})


## output to pdf file
pdf(file = "portfolios.pdf", width = 7, height = 5)

## plot the tangency and minimal variace portfolio, on efficient portfolio
par(mar = c(2.5, 2.5, 0, 0), mgp = c(1.5, 0.5, 0)) # remove margin

plot(sqrt(var_eff), retn_eff, type = "l",
     xlab = expression(sigma[p]), ylab = expression(mu[p]))
points(seVec, retnVec, col = c("red", "blue"), pch = 20)
abline(b = (retn_tan - rf) / se_tan, a = rf, lty = 2, col = "gray", lwd = 1.5)
legend("topleft", col = c("black", "gray", "red", "blue"),
       pch = c(NA, NA, 20, 20), lty = c(1, 2, NA, NA), lwd = c(1, 1.5, NA, NA),
       legend = c("Efficient Frontier", "Captial Market Line (CML)",
                  "Minimal Variance Portfolio", "Tangency Portfolio"))


## Question 2. Trade-off portfolio ---------------------------------------------
tuning <- 1e3                           # set the risk-aversion coefficient
res_trade <- solve.QP(Dmat = tuning * Q, dvec = 2 * mu,
                      Amat = matrix(1, nrow = na, ncol = 1),
                      bvec = 1, meq = 1L)
(w_trade <- res_trade$solution)
(retn_trade <- crossprod(w_trade, mu))                # expected return
(se_trade <- sqrt(crossprod(w_trade, Q %*% w_trade))) # standard error
stopifnot(all.equal(sum(w_trade), 1))

## plot the portfolio on the efficient frontier
par(mar = c(2.5, 2.5, 0, 0), mgp = c(1.5, 0.5, 0))
plot(sqrt(var_eff), retn_eff, type = "l",
     xlab = expression(sigma[p]), ylab = expression(mu[p]))
points(se_trade, retn_trade, col = "purple", pch = 20)
legend("topleft", col = c("black", "purple"), pch = c(NA, 20), lty = c(1, NA),
       lwd = c(1, NA), legend = c("Efficient Frontier", "Trade-off Portfolio"))



## Question 3. Tracking portfolio ----------------------------------------------
## set the resulting trade-off portfolio to be the benchmark portfolio
wsup <- 0.1                             # maximum holding size
winf <- 0                               # minimum holding size
Amat <- cbind(1, - diag(na), diag(na))
bvec <- c(1, rep(- wsup, na), rep(winf, na))
res_track <- solve.QP(Dmat = Q, dvec = crossprod(w_trade, Q),
                      Amat = Amat, bvec = bvec, meq = 1L)
(w_track <- res_track$solution)
(retn_track <- crossprod(w_track, mu))                # expected return
(se_track <- sqrt(crossprod(w_track, Q %*% w_track))) # standard error
stopifnot(all.equal(sum(w_track), 1))

## plot the portfolio on the efficient frontier
par(mar = c(2.5, 2.5, 0, 0), mgp = c(1.5, 0.5, 0))
plot(sqrt(var_eff), retn_eff, type = "l",
     xlab = expression(sigma[p]), ylab = expression(mu[p]))
points(se_trade, retn_trade, col = "purple", pch = 20)
points(se_track, retn_track, col = "green", pch = 20)
legend("topleft", col = c("black", "purple", "green"), pch = c(NA, 20, 20),
       lty = c(1, NA, NA), lwd = c(1, NA, NA),
       legend = c("Efficient Frontier", "Benchmark Portfolio",
                  "Tracking Portfolio"))

## output
dev.off()

File: prog/hw5/python/hw5.py
#!/usr/bin/python3


import numpy as np
import random
import cvxopt as opt
from cvxopt import matrix, solvers, blas
import matplotlib.pyplot as plt

# 1.a
######################################################
# create random returns
na = 20;  # number of assets
ns = 60;  # number of observations
random.seed(1209)
R = 0.005 + np.random.randn(na,ns)*0.015;

sigma = matrix(np.cov(R))
mu = np.matrix(np.array(np.mean(R,1)))
rf = 0.0001
mu2 = mu - rf

# set up matrices
P = sigma
q = matrix(np.zeros((na, 1)))
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(mu2)
b = matrix(1.0)

sol1 = solvers.qp(P, q, G, h, A, b)
# rescale variables to obtain weights
w1 = matrix(sol1['x']/sum(sol1['x']))
print(w1)
tanret = blas.dot(matrix(mu), w1)
tansd = np.sqrt(blas.dot(w1, sigma*w1))

# 1.b
######################################################
# Calculate mininum variance weights using quadratic programming
P = 2.0 * sigma
q = matrix(np.zeros((na, 1)))
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(np.ones((1,na)))
b = matrix(1.0)
sol2 = solvers.qp(P, q, G, h, A, b)
w2 = matrix(sol2['x'])
minret = blas.dot(matrix(mu), w2)
minsd = np.sqrt(blas.dot(w2, sigma*w2))

# Calculate efficient frontier weights using quadratic programming
N = 160
mus = [10**(8.0 * t/N) for t in range(50, N)]
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(np.ones((1,na)))
b = matrix(1.0)

# Calculate efficient frontier weights using quadratic programming
portfolios = [solvers.qp(m*sigma, -opt.matrix(mu.T), G, h, A, b)['x']
              for m in mus]
returns = [blas.dot(matrix(mu), x) for x in portfolios]
risks = [np.sqrt(blas.dot(x, sigma*x)) for x in portfolios]

# Calculate CML using tangency portfolio
slope = np.divide((tanret - rf), tansd)
cml = rf + np.array(risks) * slope

# plot
plt.ylabel('mean')
plt.xlabel('std')
line1, = plt.plot(risks, returns, 'b', label='efficient frontier')
line2, = plt.plot(minsd, minret, 'bo', label="minimum variance")
line3, = plt.plot(tansd, tanret, 'ro', label='tangency')
line4, = plt.plot(risks, cml, 'r--', label='CML')
plt.legend(handles=[line1, line2, line3, line4], loc = 2)
plt.show()

# 2.a
######################################################
lamba = 100.0
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(np.ones((1,na)))
b = matrix(1.0)

# Calculate trade-off weights using quadratic programming
sol3 = solvers.qp(lamba*P, -opt.matrix(mu.T), G, h, A, b)
w3 = matrix(sol3['x'])
traderet = blas.dot(matrix(mu), w3)
tradesd = np.sqrt(blas.dot(w3, sigma*w3))

# 2.b
######################################################
N = 160
mus = [10**(5.0 * t/N) for t in range(50, N)]
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(np.ones((1,na)))
b = matrix(1.0)

# Calculate efficient frontier weights using quadratic programming
portfolios = [solvers.qp(m*sigma, -opt.matrix(mu.T), G, h, A, b)['x']
              for m in mus]
returns = [blas.dot(matrix(mu), x) for x in portfolios]
risks = [np.sqrt(blas.dot(x, sigma*x)) for x in portfolios]

plt.ylabel('mean')
plt.xlabel('std')
line1, = plt.plot(tradesd, traderet, 'bo', label='Trade-off with $\lambda=100.0$')
line2, = plt.plot(risks, returns, 'b', label="efficient frontier")
plt.legend(handles=[line1, line2], loc = 2)
plt.show()

# 3.a
######################################################
wbm = w3
benchret = blas.dot(matrix(mu), w3)
benchsd = np.sqrt(blas.dot(w3, sigma*w3))
P = 2.0 * sigma
# Now compute covariance of individual returns and r'*wbm
rbm = np.dot(R.T, wbm)
# rbm is ns x 1 vector of portfolio returns by the benchmark portfolio (ns observations)
q = matrix(-(2.0/(ns-1)) * (np.dot(R - np.tile(np.mean(R, 1), (ns, 1)).T, \
           rbm - np.mean(rbm))))
wmin = np.random.rand(na, 1)
wmin = wmin / np.sum(wmin)
wmax = wmin + np.multiply((1 - wmin), np.random.rand(na, 1))
G = matrix(np.concatenate(( \
             -np.identity(na), \
             np.identity(na)), 0))

h = matrix(np.concatenate(( \
              -wmin, \
              wmax), 0))
A = matrix(np.ones((1,na)))
b = matrix(1.0)

sol4 = solvers.qp(P, q, G, h, A, b)
w4 = matrix(sol4['x'])
trackret = blas.dot(matrix(mu), w4)
tracksd = np.sqrt(blas.dot(w4, sigma*w4))

# 3.b
######################################################
N = 160
mus = [10**(5.0 * t/N) for t in range(50, N)]
G = matrix(-np.zeros((na,na)))
h = matrix(-np.zeros((na,1)))
A = matrix(np.ones((1,na)))
b = matrix(1.0)

# Calculate efficient frontier weights using quadratic programming
portfolios = [solvers.qp(m*sigma, -opt.matrix(mu.T), G, h, A, b)['x']
              for m in mus]
returns = [blas.dot(matrix(mu), x) for x in portfolios]
risks = [np.sqrt(blas.dot(x, sigma*x)) for x in portfolios]

plt.ylabel('mean')
plt.xlabel('std')
line1, = plt.plot(tracksd, trackret, 'go', label='tracking portfolio')
line2, = plt.plot(risks, returns, 'b', label="efficient frontier")
line3, = plt.plot(benchsd, benchret, 'ro', label="benchmark portolio")
plt.legend(handles=[line1, line2, line3], loc = 2)
plt.show()
